activate.go

```diff
@@ -615,65 +703,236 @@ func (m *activateOneContainerStateMachine) runNextState(ctx context.Context) err
+//     PlatformProtectedByStorageContainer flag (the "plainkey" platform is registered with
+//     this), as these keyslots are protected using a key that must be stored inside the
+//     first unlocked storage container. If unlocking succeeds using a keyslot protected by
+//     a platform registered with this flag, then this is sufficient to determine that the
+//     storage container is part of the same install as the first container. The recovered
+//     primary key can be used to check the binding of subsequent storage containers using
+//     the first method.
+//
+// If more than one container is unlocked using a recovery key, then all containers must be
+// unlocked using a recovery key. This is because it is only possible to prove the binding
+// of storage containers to a single container using a platform that is registered with the
+// PlatformProtectedByStorageContainer flag.
+//
+// Once any storage containers have been unlocked using a platform keyslot, all subsequent
+// containers must be unlocked using platform keyslots. This is because it is not possible
+// to properly verify that subsequent containers are bound to existing containers if they
+// are unlocked with a recovery key. In this case, existing containers must be deactivated
+// and then activated only with recovery keys.
+// XXX: A mechanism to handle this case will be provided in the future.
 func (c *ActivateContext) ActivateContainer(ctx context.Context, container StorageContainer, opts ...ActivateOption) error {
-	// Procss options. These apply on top of those supplied to NewActivateContext.
+	// Process options. These apply on top of those supplied to NewActivateContext.
 	cfg := c.cfg.Clone()
 	for _, opt := range opts {
 		opt.ApplyOptionToConfig(cfg)
 	}
 
-	// TODO: When we retain activation state, make sure we obtain the activation
-	// status and keyslot errors from the state machine before returning.
-
 	// Create and run a state machine for this activation
-	sm := newActivateOneContainerStateMachine(container, cfg)
+	var flags activateOneContainerStateMachineFlags
+	switch {
+	case c.state.TotalActivatedContainers() == 0:
+		// If this is the first container to be activated, permit either
+		// a platform or recovery keyslot to be used.
+		flags = activatePermitPlatformKey | activatePermitRecoveryKey
+	case c.state.TotalActivatedContainers() == 1 && c.state.NumActivatedContainersWithRecoveryKey() == 1:
+		// If this is the second container to be activated and the first
+		// was activated with a recovery keyslot, permit either a platform
+		// or recovery keyslot to be used. The platform keyslot must be
+		// protected with a platform registered with the
+		// PlatformProtectedByStorageContainer flag.
+		flags = activatePermitPlatformKey | activatePermitRecoveryKey | activateRequirePlatformKeyProtectedByStorageContainer
+	case c.state.NumActivatedContainersWithRecoveryKey() > 1:
+		// If more than one storage container has been unlocked with a
```

Comment:
What if we had 3 containers. 2 unlocked with recovery keys, but they both have matching platform keys. Which means we unlock the 3rd one with plainkey. In this case the primary key of the 3rd one should match the 2 first containers, right?

Should it be that we remember which containers where we found platform keys. And once we unlock a disk with a plainkey, we consider any recovery unlocked containers containing the platform key to have a matching primary key as the plainkey unlocked one.

I know it is not a scenario we have for Ubuntu Core. But it makes more sense. Or at least if this is correct, I understood the intention here.